keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2052 @@
+(https://github.com/kubernetes/enhancements/pull/1703).  Each “fake”
+CSI driver would have to implement and install a scheduler extender
+because storage capacity tracking only considers volume size as
+criteria for selecting nodes, which is not applicable for custom
+resources.
+
+### Extend volume support
+
+The StorageClass and PersistentVolumeClaim structs could be extended
+to allow custom parameters. Together with an extension of the CSI
+standard that would address the main objection against the previous
+alternative.
+
+However, SIG-Storage and the CSI community would have to agree to this
+kind of reuse and accept that some of the code maintained by them
+becomes more complex because of these new use cases.
+
+### Extend Device Plugins
+
+The device plugins API could be extended to implement some of the
+requirements mentioned in the “Motivation” section of this
+document. There were certain attempts to do it, for example an attempt
+to [add ‘Deallocate’ API call](https://github.com/kubernetes/enhancements/pull/1949) and [pass pod annotations to 'Allocate' API call](https://github.com/kubernetes/kubernetes/pull/61775)
+
+However, most of the requirements couldn’t be satisfied using this
+approach as they would require major incompatible changes in the
+Device Plugins API. For example: partial and optional resource
+allocation couldn’t be done without changing the way resources are
+currently declared on the Pod and Device Plugin level.
+
+Extending the device plugins API to use [Container Device Interface](https://github.com/container-orchestrated-devices/container-device-interface)
+would help address some of the requirements, but not all of them.
+
+NodePrepareResource and NodeUnprepareResource could be added to the Device Plugins API and only get called for
+resource claims.
+
+However, this would mean that
+developers of the device plugins would have to implement mandatory
+API calls (ListAndWatch, Allocate), which could create confusion
+as those calls are meaningless for the Dynamic Resource Allocation
+purposes.
+
+Even worse, existing device plugins would have to implement the new
+calls with stubs that return errors because the generated Go interface
+will require them.
+
+It should be also taken into account that device plugins API is
+beta. Introducing incompatible changes to it may not be accepted by
+the Kubernetes community.
+
+### Webhooks instead of ResourceClaim updates
+
+In the current design, scheduler and the third-party resource driver communicate by
+updating fields in a ResourceClaim. This has several advantages compared to an
+approach were kube-scheduler retrieves information from the resource driver
+via HTTP:
+* No need for a new webhook API.
+* Simpler deployment of a resource driver because all it needs are
+  credentials to communicate with the apiserver.
+* Current status can be checked by querying the ResourceClaim.
+
+The downside is higher load on the apiserver and an increase of the size of
```

Comment:
CRDs versus webhooks have two competing benefits, respectively:
- easier installation
- faster scheduling

Do you have any sort of supporting evidence that users prefer easier installation over faster scheduling?

Reply:
I know that some users don't have that choice. For example, recent OpenShift removed the API for configuring scheduler extenders in the scheduler.

From users of PMEM-CSI I know that they skipped installing the scheduler extender because they found that part too complicated and it wasn't needed for simple trials. Luckily, storage capacity tracking will soon be available on all supported Kubernetes releases, so the extender will no longer be needed to ensure correct pod scheduling for pods with local volumes.

It is part of the plan to gather feedback from users. But asking potential users now seems premature: we first need to determine how fast scheduling is for real workloads before we can ask whether that is fast enough.

Reply:
> I know that some users don't have that choice. For example, recent OpenShift removed the API for configuring scheduler extenders in the scheduler.

That is true for the generic webhook API that we have. But what if we introduce a better thought API that is more like CRDs webhooks in the sense that you can register them dynamically.

> It is part of the plan to gather feedback from users.

If we decided later on that webhooks is a better answer from this feedback, are there any new fields in the PodSpec that we would regret?

I suppose changing fields in ResourceClaim is fine, as they would first be introduced as v1alpha1

Reply:
> But what if we introduce a better thought API that is more like CRDs webhooks in the sense that you can register them dynamically.

That implies that kube-scheduler must be deployed so that it actually can connect to a service inside the cluster. This is not currently required and I don't know whether it is something that should be made a requirement for all Kubernetes installations.

> If we decided later on that webhooks is a better answer from this feedback, are there any new fields in the PodSpec that we would regret?

All of these  fields will be treated as alpha (not enabled without the feature gate, can be changed from one release to the next). Of course it would be nicer to get it right on the first try, but if not, then we can still change them.

Reply:
In the revised API intro I explicitly call out the alpha state of the new fields.